add offline batch ingestion tech doc (#8251)
opensearch-project/documentation-website: _ml-commons-plugin/remote-models/async-batch-ingestion.md

---
layout: default
title: Asynchronous batch ingestion
nav_order: 90
parent: Connecting to externally hosted models 
grand_parent: Integrating ML models
---

# Asynchronous batch ingestion
**Introduced 2.17**
{: .label .label-purple }

[Batch ingestion]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/batch-ingestion/) configures an ingest pipeline, which processes documents one by one. For each document, batch ingestion calls an externally hosted model to generate text embeddings from the document text and then ingests the document, including text and embeddings, into an OpenSearch index.

> An alternative to this real-time process, _asynchronous_ batch ingestion, ingests both documents and their embeddings generated outside of OpenSearch and stored in a remote file server, such as Amazon S3 or OpenAI. Asynchronous ingestion returns a task ID and runs asynchronously to ingest data into your k-NN cluster for neural search offline. You can use asynchronous batch ingestion together with [Batch Predict API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/model-apis/batch-predict/) to perform inference asynchronously. Batch predict operation takes an input file, which contains documents, and calls an externally hosted model to generate embeddings for those documents into an output file. You can then use asynchronous batch ingestion to ingest both the input file containing documents and the output file containing their embeddings into an OpenSearch index.

Reviewer: ```suggestion
An alternative to this real-time process, _asynchronous_ batch ingestion, ingests both documents and their embeddings generated outside of OpenSearch and stored on a remote file server, such as Amazon Simple Storage Service (Amazon S3) or OpenAI. Asynchronous ingestion returns a task ID and runs asynchronously to ingest data into your k-NN cluster for offline neural search. You can use asynchronous batch ingestion together with the [Batch Predict API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/model-apis/batch-predict/) to perform inference asynchronously. The batch predict operation takes an input file containing documents and calls an externally hosted model to generate embeddings for those documents in an output file. You can then use asynchronous batch ingestion to ingest both the input file containing documents and the output file containing their embeddings into an OpenSearch index.
```

> As of OpenSearch 2.17, Batch Ingestion API is verified to work with Amazon SageMaker, Amazon Bedrock, and OpenAI.

Reviewer: ```suggestion
As of OpenSearch 2.17, the Batch Ingestion API is supported by Amazon SageMaker, Amazon Bedrock, and OpenAI.
```

{: .note}

## Prerequisites

Before using asynchronous batch ingestion, you must generate text embeddings using a model of your choice and store the output on a file server, such as Amazon S3. For example, you can store the output of a Batch API call to an Amazon SageMaker text embedding model in a file with the Amazon S3 output path `s3://offlinebatch/output/sagemaker_batch.json.out`. The output is in JSONL format, with each line representing a text embedding result. The file contents have the following format:

```
{"SageMakerOutput":[[-0.017166402,0.055771016,...],[-0.06422759,-0.004301484,...],"content":["this is chapter 1","harry potter"],"id":1}
{"SageMakerOutput":[[-0.017455402,0.023771016,...],[-0.02322759,-0.009101284,...],"content":["this is chapter 2","draco malfoy"],"id":1}
...
```

## Ingesting data from a single file

> First, create a k-NN index where you'll ingest the data. The fields in the k-NN index represent the structure of the data in the source file. 

Reviewer: ```suggestion
First, create a k-NN index into which you'll ingest the data. The fields in the k-NN index represent the structure of the data in the source file. 
```

> In this example, the source file contains documents containing titles and chapters, along with their corresponding embeddings. Thus, you'll create a k-NN index with fields `id`, `chapter_embedding`, `chapter`, `title_embedding`, and `title`:

Reviewer: ```suggestion
In this example, the source file holds documents containing titles and chapters, along with their corresponding embeddings. Thus, you'll create a k-NN index with the fields `id`, `chapter_embedding`, `chapter`, `title_embedding`, and `title`:
```

```json
PUT /my-nlp-index
{
  "settings": {
    "index.knn": true
  },
  "mappings": {
    "properties": {
      "id": {
        "type": "text"
      },
      "chapter_embedding": {
        "type": "knn_vector",
        "dimension": 384,
        "method": {
          "engine": "nmslib",
          "space_type": "cosinesimil",
          "name": "hnsw",
          "parameters": {
            "ef_construction": 512,
            "m": 16
          }
        }
      },
      "chapter": {
        "type": "text"
      },
      "title_embedding": {
        "type": "knn_vector",
        "dimension": 384,
        "method": {
          "engine": "nmslib",
          "space_type": "cosinesimil",
          "name": "hnsw",
          "parameters": {
            "ef_construction": 512,
            "m": 16
          }
        }
      },
      "title": {
        "type": "text"
      }
    }
  }
}
```
{% include copy-curl.html %}

> When using an S3 file as the source for asynchronous batch ingestion, you must map the fields in the source file to fields in the index in order to indicate where each piece of data is ingested. If no JSON path is provided for a field, that field in the k-NN index will be set to `null`.

Reviewer: ```suggestion
When using an S3 file as the source for asynchronous batch ingestion, you must map the fields in the source file to fields in the index in order to indicate into which index each piece of data is ingested. If no JSON path is provided for a field, that field will be set to `null` in the k-NN index.
```

> In the `field_map`, provide the location where the data for each field can be found in the source file. You can also specify fields to be ingested directly into your index without making any changes to the source file by adding their JSON path to the `ingest_fields` array. For example, in the following asynchronous batch ingestion request, the element with the JSON path `$.id` from the source file is ingested directly into the `id` field of your index. To ingest this data from the Amazon S3 file, send the following request to your OpenSearch endpoint:

Reviewer: ```suggestion
In the `field_map`, indicate the location of the data for each field in the source file. You can also specify fields to be ingested directly into your index without making any changes to the source file by adding their JSON paths to the `ingest_fields` array. For example, in the following asynchronous batch ingestion request, the element with the JSON path `$.id` from the source file is ingested directly into the `id` field of your index. To ingest this data from the Amazon S3 file, send the following request to your OpenSearch endpoint:
```

> 

Reviewer: Above, third sentence: "the source file element with the JSON path `$.id` is ingested..."?
